Dual-stack Support (#202)
metal-stack/docs: docs/src/development/proposals/MEP13/README.md

# IPv6 Support

IPv6 support is required to be able to create Kubernetes clusters with either IPv6 single- or dual-stack enabled.
With immanent shortage of IPv4 addresses the need to be able to use IPv6 increased.

Full IPv6 dual-stack Support was added to Kubernetes with v1.23 as stable.

Gardener on the other hand does not yet have full IPv6 dual-stack support. See: https://github.com/gardener/gardener/issues/7051

> Metal Stack manages Cidrs and IP Addresses with the [go-ipam](https://github.com/metal-stack/go-ipam) library, which gained full IPv6 Support already in 2021 https://metal-stack.io/blog/2021/02/ipv6-part1.

Reviewer: ```suggestion
metal-stack manages CIDRs and IP addresses with the [go-ipam](https://github.com/metal-stack/go-ipam) library, which gained full IPv6 Support already in 2021 (see https://metal-stack.io/blog/2021/02/ipv6-part1).
```

But this was only the foundation, to get full IPv6 support for all aspects which are managed by metal-stack.io, further work needs to be done.

## General Decisions

### Networks

> Currently metal-stack organizes Cidrs/Prefixes in a `network`. A network can consist of multiple cidrs from the same address family,
> for example if you need internet connectivity, you start with a small cidrs but with time you will need more ipaddresses and by more internet cidrs.
> Then you simply add these additional cidrs you bought, to the already existing internet `network` and more servers can be created with internet addresses.

Reviewer: ```suggestion
Currently, metal-stack organizes CIDRs/prefixes into a `network' resource in the metal-api. A network can consist of multiple CIDRs from the same address family. For example, if an operator wants to provide Internet connectivity to provisioned machines, they can start with a small network CIDRs. The number of managed network prefixes can then be expanded as needed over time.
```

With IPv6 we have to choose between two options:

#### Network per Address Family

> This means that we allow networks with cidrs from one address family only, one for IPv4 and one for IPv6

Reviewer: ```suggestion
This means that we allow networks with CIDRs from one address family only, one for IPv4 and one for IPv6.
```
